src/learner.cc

```diff
@@ -187,9 +190,13 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
  */
 class LearnerImpl : public Learner {
  public:
-  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> >  cache)
-      : need_configuration_{true}, cache_(std::move(cache)) {
+  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> > cache)
+      : need_configuration_{true},
+        cache_{new PredictionContainer} {
     monitor_.Init("Learner");
+    for (std::shared_ptr<DMatrix> const& d : cache) {
+      cache_->Cache(d, GenericParameter::kCpuId);
```

Comment:
So the shared pointer doesn't get copied here?

Reply:
@RAMitchell We don't keep the shared pointer in booster after this PR.  This way the DMatrix object can be destructed without first deleting the booster.